File: src/p2.hs
--	tail-recursive fibonacci
f _ v 1 = v
f a b n = f b (a+b) (n-1)

trfib n = f 0 1 n

--	infinite-list fibonacci
inffib = zipWith (+) (0:1:inffib) (1:inffib)
res up = sum $ filter even $ foldr (\e l -> if (e < up) then e:l else []) [] inffib

File: src/p2.jl
using Lazy

curry(f, x) = (xs...) -> f(x, xs...)

fibs = @lazy 0:1:(fibs + tail(fibs))

under = curry(takewhile, (x -> x <= 4e6))
even = curry(filter, iseven)
solve = (sum ∘ even ∘ under)

println(solve(fibs))
